src/createDynamicHostResolver.js

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ ****************************************************************************************/
+
+var window = require('@adobe/reactor-window');
+
+module.exports = function (turbineEmbedCode, cdnAllowList, debugController) {
+  // even an empty list is flagging to us that we're trying to enforce dynamic
+  var isDynamicEnforced = Array.isArray(cdnAllowList);
+  var embedPattern = /^https?:\/\/.*/;
+
+  // TODO: web only? I think embedded TVs wouldn't have
+  //  __satellite.container.dynamicEnforced turned on
+  var turbineUrl = document.createElement('a');
+  turbineUrl.href = turbineEmbedCode;
+  if (
+    (!embedPattern.test(turbineEmbedCode) || !turbineUrl.host) &&
+    isDynamicEnforced === true
+  ) {
+    var missingEmbedCodeError = new Error(
+      'Unable to find the Library Embed Code for Dynamic Host Resolution.'
+    );
+    missingEmbedCodeError.code = 'dynamic_host_resolver_constructor_error';
+    throw missingEmbedCodeError;
+  }
+
+  if (isDynamicEnforced && cdnAllowList.indexOf(turbineUrl.hostname) === -1) {
+    var dynamicDeniedError = new Error(
+      'This library is not authorized for this domain. ' +
+        'Please contact your CSM for more information.'
+    );
+    dynamicDeniedError.code = 'dynamic_host_not_allowed';
+    throw dynamicDeniedError;
+  }
+
+  var shouldAugment = Boolean(isDynamicEnforced && turbineUrl);
+
+  /**
+   * Returns the host of the Turbine embed code, or an empty string if Dynamic Host
+   * is not enabled.
+   * @returns {string}
+   */
+  var getTurbineHost = function () {
+    if (shouldAugment) {
+      // be sure we always force https to Adobe managed domains.
+      // IE 10/11 returns the :443 protocol when modern browsers don't, so this replacement
+      // is bringing every browser in line with the same return value
+      var sanitizedHost = turbineUrl.host;
+      if (/:80$/.test(sanitizedHost)) {
+        sanitizedHost = sanitizedHost.replace(':80', '');
+      } else if (/:80\/$/.test(sanitizedHost)) {
+        sanitizedHost = sanitizedHost.replace(':80/', '');
+      } else if (/:443$/.test(sanitizedHost)) {
+        sanitizedHost = sanitizedHost.replace(':443', '');
+      } else if (/:443\/$/.test(sanitizedHost)) {
+        sanitizedHost = sanitizedHost.replace(':443/', '');
+      }
+
+      return 'https://' + sanitizedHost;
+    }
+
+    return '';
+  };
+
+  /**
+   * Returns a url decorated with the host of the Turbine embed code. If Dynamic host
+   * is disabled, the original sourceUrl is returned unmodified.
+   * @param sourceUrl
+   * @returns {string|*}
+   */
+  var decorateWithDynamicHost = function (sourceUrl) {
+    if (shouldAugment) {
```

Comment:
I think we should test that sourceUrl is also a string.

Reply:
So down below where I was saying that Forge has built me a list of filePaths, it seems like I should be able to rely on that contract and if there's something wrong then blow up here when we go to decorate a non-string. What do you think @spencerroan ?

Reply:
I think we should not throw an error either here or in traverseDelegateProperties.js. Maybe we could log a warning.